rules_daml/daml.bzl

```
@@ -368,6 +369,79 @@ def daml_compile_with_dalf(
         dalf = name + ".dalf",
     )
 
+def generate_and_track_dar_hash_file(name):
+    """
+    Given a 'name', defines the following targets:
+      (1) ':{name}-dar-hash-file-matches',
+      (2) ':{name}-golden-hash'.
+      (3) ':{name}-generated-hash',
+    Target (1) is a test that checks that the files in targets (2) and (3) match.
+    It can also replace (2) with the output of (3) if run with the flag '--accept'.
+    Target (2) is an indirection to the file
+      ':{name}.dar-hash',
+    which must be present in the package where this macro is used.
+    Target (3) is transitively defined by 'generate_dar_hash_file'.
+    """
+    generate_dar_hash_file(name)
+
+    native.filegroup(
+        name = name + "-golden-hash",
+        srcs = native.glob([name + ".dar-hash"]),
+        visibility = ["//visibility:public"],
+    )
+
+    test_name = name + "-dar-hash-file-matches"
+
+    lbl = "//{package}:{target}".format(
+        package = native.package_name(),
+        target = test_name,
+    )
+
+    native.sh_test(
+        name = test_name,
+        srcs = ["//bazel_tools:match-golden-file"],
+        args = [
+            lbl,
+            "$(location :{}-generated-hash)".format(name),
+            "$(location :{}-golden-hash)".format(name),
+            "$(POSIX_DIFF)",
+        ],
+        data = [
+            ":{}-generated-hash".format(name),
+            ":{}-golden-hash".format(name),
+        ],
+        toolchains = [
+            "@rules_sh//sh/posix:make_variables",
+        ],
+        deps = [
+            "@bazel_tools//tools/bash/runfiles",
+        ],
+    )
+
+def generate_dar_hash_file(name):
+    """
+    Given a 'name', defines a target of the form
+      ':{name}-generated-hash'.
+    with a single file of the form
+      ':{name}-generated.dar-hash'.
+    The resulting file will contain one line per file present in
+      ':{name}.dar',
+    sorted by filename and excluding *.hi and *.hie files, where each line
+    has the sha256sum for the file followed by the path of the file relative to
+    the root of the dar.
+    """
+    native.genrule(
+        name = name + "-generated-hash",
+        srcs = ["//rules_daml:generate-dar-hash", ":{}.dar".format(name)],
+        tools = ["@python_dev_env//:python"] if is_windows else [],
+        outs = [name + "-generated.dar-hash"],
+        cmd = "{python} {exe} {dar} > $@".format(
+            python = "$(execpath @python_dev_env//:python)" if is_windows else "python",
+            exe = "$(location //rules_daml:generate-dar-hash)".format(name),
```

What is the `.format(name)` bit doing here?

---

ah yes the constant function 🤦 thanks for spotting it

```suggestion
            exe = "$(location //rules_daml:generate-dar-hash)",
```